pkg/containermeta/store.go

```diff
@@ -0,0 +1,282 @@
+
+	// this is a `ch []Event` instead of a `ch Event` to improve
+	// throughput, as bursts of events are as likely to occur as isolated
+	// events, especially at startup or with collectors that periodically
+	// pull changes.
+	ch := make(chan []Event, bufferSize)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.subscribers[ch] = struct{}{}
+
+	// TODO(juliogreff): do the initial notifying of existing entities
+
+	return ch
+}
+
+// Unsubscribe ends a subscription to entity events and closes its channel.
+func (s *Store) Unsubscribe(ch chan []Event) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	delete(s.subscribers, ch)
+	close(ch)
+}
+
+// GetContainer returns metadata about a container.
+func (s *Store) GetContainer(id string) (Container, error) {
+	s.mu.RLock()
+	defer s.mu.Unlock()
+
+	c, ok := s.containers[id]
+	if !ok {
+		// TODO(juliogreff): create a NotFound error type
+		return c, fmt.Errorf("container %q not found", id)
+	}
+
+	return c, nil
+}
+
+// GetKubernetesPod returns metadata about a Kubernetes pod.
+func (s *Store) GetKubernetesPod(id string) (KubernetesPod, error) {
+	s.mu.RLock()
+	defer s.mu.Unlock()
+
+	p, ok := s.kubePods[id]
+	if !ok {
+		// TODO(juliogreff): create a NotFound error type
+		return p, fmt.Errorf("kubernetes pod %q not found", id)
+	}
+
+	return p, nil
+}
+
+// GetECSTask returns metadata about an ECS task.
+func (s *Store) GetECSTask(id string) (ECSTask, error) {
+	s.mu.RLock()
+	defer s.mu.Unlock()
+
+	t, ok := s.ecsTasks[id]
+	if !ok {
+		// TODO(juliogreff): create a NotFound error type
+		return t, fmt.Errorf("container %q not found", id)
+	}
+
+	return t, nil
+}
+
+func (s *Store) startCandidates(ctx context.Context) {
+	for id, c := range s.candidates {
+		err := c.Start(ctx, s.eventCh)
+
+		// Leave candidates that returned a retriable error to be
+		// re-started in the next tick
+		if err != nil && retry.IsErrWillRetry(err) {
+			continue
+		}
+
+		// Store successfully started collectors for future reference
+		if err == nil {
+			s.collectors = append(s.collectors, c)
+		}
+
+		// Remove non-retriable and successfully started collectors
+		// from the list of candidates so they're not retried in the
+		// next tick
+		delete(s.candidates, id)
+	}
+}
+
+func (s *Store) pull(ctx context.Context) {
+	// NOTE: s.collectors is not guarded by a mutex as it's only called by
+	// the store itself, and the store runs on a single goroutine. If this
+	// method is made public in the future, we need to guard it.
+	for _, c := range s.collectors {
+		// Run each pull in its own separate goroutine to reduce
+		// latency and unlock the main goroutine to do other work.
+		go func(c Collector) {
+			err := c.Pull(ctx)
+			if err != nil {
+				log.Warnf("error pulling from collector: %s", err.Error())
+			}
+		}(c)
+	}
+}
+
+func (s *Store) handleEvents(evs []Event) {
+	for _, ev := range evs {
+		meta := ev.Entity.GetMeta()
+
+		switch meta.Kind {
+		case KindContainer:
+			s.handleContainerEvent(ev)
```

Comment:
How are we going to handle receiving the same container from different sources? (kubelet and docker for example)

Reply:
The merging is going to be done in the store itself, so to the outside world there is a single source. Still an open question on how to do it (especially how to handle priority between collectors and what data can overwrite each other), but at least the complexity is centralized instead of spread over whatever needs to use the data.